### PR TITLE
Do not use native obj Jinja2 env to render error emails

### PR DESCRIPTION
**Pull request by andyfcx**
https://github.com/apache/airflow/issues/22152

I added a flag **force_string** to make sure returning will be string,
Testing code still in progress.
Need more discussion.

---

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/abstractoperator.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ def get_template_env(self) -> "jinja2.Environment":
 
         dag = self.get_dag()
         if dag:
-            return dag.get_template_env()
+            return dag.get_template_env(force_string=True)
```
Comment (uranusjr) [inline airflow/models/abstractoperator.py:118]: When is this False?

Comment (andyfcx) [reply airflow/models/abstractoperator.py:118]: Default as False

**Comment by uranusjr** [comment]
(Some context for this) I’m helping @andyfcx to work out a solution to #22152.

The main problem with #22152 is that, when you set `render_template_as_native_obj` on a DAG, it would use a `NativeEnvironment` to create a Jinja2 template object, which in turn will be able to be rendered into a native object. However, the email-sending mechanism also uses the same Jinja2 environment, but this environment cannot correctly create a template to render into the email’s content (which must be a string, obviously).

So the proposed solution here is to add an additional flag on `get_template_env` to control what environment to return. If `force_string` is True, `get_template_env` will always create a `SandboxedEnvironment` (which is rendered into a string), even if `render_template_as_native_obj` is True. A `NativeEnvironment` is used only if `render_template_as_native_obj` is True _and_ `force_string` is False.

But the implementation in the PR still needs some work. I’m also not entirely convinced this is the best solution, and any suggestions are welcomed.

File: airflow/models/abstractoperator.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/abstractoperator.py:118]: Oh, right 😓

Comment (uranusjr) [reply airflow/models/abstractoperator.py:118]: I don’t think this is correct though. This function is also used when the task is being executed (in `render_template_fields`), and in _that_ situation we want to use `NativeEnvironment` if `render_template_as_native_obj` is True. so this `get_template_env` also need to have a `force_string` argument, and only `get_email_subject_content` should set it to True.

File: airflow/models/dag.py [hunk 2]
```diff
@@ -1278,7 +1278,10 @@ def get_template_env(self) -> jinja2.Environment:
             jinja_env_options.update(self.jinja_environment_kwargs)
         env: jinja2.Environment
         if self.render_template_as_native_obj:
-            env = airflow.templates.NativeEnvironment(**jinja_env_options)
+            if force_string:
+                env = airflow.templates.SandboxedEnvironment(**jinja_env_options)
+            else:
+                env = airflow.templates.NativeEnvironment(**jinja_env_options)
         else:
             env = airflow.templates.SandboxedEnvironment(**jinja_env_options)
```
Comment (uranusjr) [inline airflow/models/dag.py]: Can be shortened to

```python
use_native_env = self.render_template_as_native_obj and not force_string
if use_native_env:
    env = airflow.templates.NativeEnvironment(**jinja_env_options)
else:
    env = airflow.templates.SandboxedEnvironment(**jinja_env_options)
```

Comment (andyfcx) [reply airflow/models/dag.py]: done

**Pull request closed without merging**